mis/solver/greedyMIS.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+import copy
+from typing import Dict, List, Tuple
+
+import networkx as nx
+
+from mis.shared.types import MISInstance, MISSolution
+from mis.pipeline.execution import Execution
+from mis.pipeline.basesolver import BaseSolver
+from mis.pipeline.config import SolverConfig
+from mis.solver.greedymapping import GreedyMapping
+from mis.solver.lattice import Lattice
+
+
+class GreedyMISSolver(BaseSolver):
+    def __init__(
+        self,
+        instance: MISInstance,
+        solver: BaseSolver,
+    ) -> None:
+        super().__init__(instance, solver.config)
+
+        self.original_instance = instance
+        self.sub_solver = solver
+        self.config: SolverConfig = solver.config
+        self.seed = 0
+
+        self.lattice = Lattice(
+            lattice_coords=self.config.lattice_coords,
+            rydberg_blockade=self.config.rydberg_blockade,
+            seed=self.seed,
+        )
+
+    def solve(self) -> Execution[List[MISSolution]]:
+        return self._solve_recursive(self.original_instance)
+
+    def _solve_recursive(self, instance: MISInstance) -> Execution[List[MISSolution]]:
+        graph = instance.graph
+        if len(graph) <= self.config.exact_solving_threshold:
+            return self._exact_mis(instance)
+
+        mappings = self._generate_subgraphs(graph)
+        best_solution: MISSolution | None = None
+
+        for mapping in mappings:
+            subgraph = graph.subgraph(mapping.keys())
+            if len(subgraph) <= self.config.exact_solving_threshold:
+                continue
+
+            lattice_subgraph = self._generate_lattice_graph(graph, mapping)
+            sub_instance = MISInstance(graph=lattice_subgraph)
+            sub_solver = type(self.sub_solver)(sub_instance, self.config)
+
+            results = sub_solver.solve().result()[:self.config.max_number_of_solutions]
+            inv_map = {v: k for k, v in mapping.items()}
+
+            for partial in results:
+                logical_nodes = [inv_map[n] for n in partial.nodes]
+                reduced_graph = self._remove_nodes(graph, logical_nodes)
+                remainder_instance = MISInstance(reduced_graph)
+
+                remainder_exec = self._solve_recursive(remainder_instance)
```

Comment:
What's the maximal recursion level in Python? Is this something that we could hit? If not, could you document it somewhere? If it is something we could hit, should we derecursify the code?